maze_transformer/training/tokenizer.py

```diff
@@ -1,66 +1,47 @@
-    CoordArray,
+    Coord,
     CoordTup,
     LatticeMaze,
 )
 
 # pylint: disable=unused-import
 
 
-@dataclass(frozen=True, kw_only=True)
-class MazeTokenizer:
-    """solved maze for serialization"""
-
-    maze: LatticeMaze
-    solution: CoordArray
-    metadata: dict = field(default_factory=dict)
-
-    pos_start = property(lambda self: self.solution[0])
-    pos_end = property(lambda self: self.solution[-1])
-
-    def as_tokens(
-        self,
-        node_token_map: dict[CoordTup, str],
-        solution: bool = True,
-    ) -> list[str]:
-        """serialize maze and solution to tokens"""
-        tokens: list[str] = [
-            # give adjacency list
-            SPECIAL_TOKENS["adjlist_start"],
-            *chain.from_iterable(
+def maze_to_tokens(
+    maze: LatticeMaze,
+    solution: list[Coord],
+    node_token_map: dict[CoordTup, str],
+) -> list[str]:
+    """serialize maze and solution to tokens"""
+    tokens: list[str] = [
+        # give adjacency list
+        SPECIAL_TOKENS["adjlist_start"],
+        *chain.from_iterable(
+            [
                 [
-                    [
-                        node_token_map[tuple(c_s.tolist())],
-                        SPECIAL_TOKENS["connector"],
-                        node_token_map[tuple(c_e.tolist())],
-                        SPECIAL_TOKENS["adjacency_endline"],
-                    ]
-                    for c_s, c_e in self.maze.as_adj_list()
+                    node_token_map[tuple(c_s.tolist())],
+                    SPECIAL_TOKENS["connector"],
+                    node_token_map[tuple(c_e.tolist())],
+                    SPECIAL_TOKENS["adjacency_endline"],
                 ]
-            ),
-            SPECIAL_TOKENS["adjlist_end"],
-            # give origin
-            SPECIAL_TOKENS["origin_start"],
-            node_token_map[tuple(self.pos_start)],
-            SPECIAL_TOKENS["origin_end"],
-            # give target
-            SPECIAL_TOKENS["target_start"],
-            node_token_map[tuple(self.pos_end)],
-            SPECIAL_TOKENS["target_end"],
-        ]
-
-        if solution:
-            # give path
-            tokens.extend(
-                [
-                    SPECIAL_TOKENS["path_start"],
-                    *[node_token_map[tuple(c.tolist())] for c in self.solution],
-                    SPECIAL_TOKENS["path_end"],
-                ]
-            )
-
-        return tokens
+                for c_s, c_e in maze.as_adj_list()
+            ]
+        ),
+        SPECIAL_TOKENS["adjlist_end"],
+        # give origin
+        SPECIAL_TOKENS["origin_start"],
+        node_token_map[tuple(solution[0])],
+        SPECIAL_TOKENS["origin_end"],
+        # give target
+        SPECIAL_TOKENS["target_start"],
+        node_token_map[tuple(solution[-1])],
+        SPECIAL_TOKENS["target_end"],
+        SPECIAL_TOKENS["path_start"],
+        *[node_token_map[tuple(c.tolist())] for c in solution],
+        SPECIAL_TOKENS["path_end"],
```

Comment:
⚠️ Will this modify the tokenization scheme such that "path_start" and "path_end" tokens are always included, regardless of whether the solution is empty? 

This might be a change for the better, but I don't think it belongs in a refactor. At the very least we should check with the team that this behaviour is desired.

Reply:
Yes, with this change solution is a required parameter and will always be included. I wouldn't call this a change to the tokenization scheme though, everything gets tokenized the same way.

This change actually has no functional effect, it just simplifies things a lot. In the [previous version](https://github.com/AISC-understanding-search/maze-transformer/blob/main/maze_transformer/training/tokenizer.py), we needed to construct a `MazeTokenizer` and then call `as_tokens`. `solution` was a required property when constructing the tokenizer, and then when calling `as_tokens` you could specify whether to include it in the result or not. But when actually using the function, we always included the solution, except in `plot_attention.ipynb`, which I've adjusted to maintain the old behaviour.

I think it's overall simpler for the tokenizer to not have too much logic and just return the full tokenization. In my evals branch I've added a bunch of [utility functions](https://github.com/AISC-understanding-search/maze-transformer/blob/99-wip-2/maze_transformer/utils/token_utils.py) to easily get different parts of the tokenization, including one that gets everything up to but not including the solution, which is equivalent to the old behaviour here when the solution is excluded.

Reply:
Big fan of this change as the behaviour w.r.t `solution` caused me some annoyance in trying to include `LatticeMazes` In the `WrappedTokenizer` encoding method